dask/array/reshape.py

```diff
@@ -364,3 +375,198 @@ def reshape(x, shape, merge_chunks=True, limit=None):
+    >>> result.chunks
+    ((3,), (4, 2, 2, 1))
+
+    The resulting chunks are calculated automatically to match the new shape.
+
+    >>> result.compute()
+    array([[ 0,  1,  3,  4,  2,  5,  6,  7,  8],
+           [ 9, 10, 12, 13, 11, 14, 15, 16, 17],
+           [18, 19, 21, 22, 20, 23, 24, 25, 26]])
+
+    >>> result = reshape_blockwise(result, (3, 3, 3), chunks=x.chunks)
+    >>> result.chunks
+    ((3,), (2, 1), (2, 1))
+
+    The resulting chunks are taken from the input. Chaining the reshape operation
+    together like this reverts the previous reshaping operation that reduces the
+    number of dimensions.
+
+    >>> result.compute()
+    array([[[ 0,  1,  2],
+            [ 3,  4,  5],
+            [ 6,  7,  8]],
+    <BLANKLINE>
+           [[ 9, 10, 11],
+            [12, 13, 14],
+            [15, 16, 17]],
+    <BLANKLINE>
+           [[18, 19, 20],
+            [21, 22, 23],
+            [24, 25, 26]]])
+    """
+    if shape in [-1, (-1,)]:
+        shape = (reduce(mul, x.shape),)
+
+    if not isinstance(shape, tuple):
+        shape = (shape,)
+
+    outname = "reshape-blockwise-" + tokenize(x, shape)
+    chunk_tuples = list(product(*(range(len(c)) for i, c in enumerate(x.chunks))))
+
+    if len(shape) > x.ndim:
+        if chunks is None:
+            raise TypeError("Need to specify chunks if expanding dimensions.")
+        out_shapes = list(product(*(c for c in chunks)))
+        out_chunk_tuples = list(product(*(range(len(c)) for c in chunks)))
+        in_shapes = list(product(*(c for c in x.chunks)))
+        non_matching_chunks = [
+            (i, in_c, out_c)
+            for i, (in_c, out_c) in enumerate(zip(in_shapes, out_shapes))
+            if math.prod(in_c) != math.prod(out_c)
+        ]
+        if non_matching_chunks:
+            raise ValueError(
+                f"Chunk sizes do not match for the following chunks: "
+                f"{[x[0] for x in non_matching_chunks[:5]]}. \n"
+                f"The corresponding chunksizes are: {[x[1:] for x in non_matching_chunks[:5]]}. "
+                f"(restricted to first 5 entries)."
+            )
+
+        dsk = {
+            (outname,)
+            + tuple(chunk_out): (_reshape_blockwise, (x.name,) + tuple(chunk_in), shape)
+            for chunk_in, chunk_out, shape in zip(
+                chunk_tuples, out_chunk_tuples, out_shapes
+            )
+        }
+
+        graph = HighLevelGraph.from_collections(outname, dsk, dependencies=[x])  # type: ignore[arg-type]
+        return Array(graph, outname, chunks, meta=x._meta)
+
+    elif len(shape) < x.ndim:
+        if chunks is not None:
+            raise ValueError(
+                "Setting chunks is not allowed when reducing the number of dimensions."
+            )
+
+    _, _, mapper_in, one_dimensions = reshape_rechunk(
+        x.shape, shape, x.chunks, disallow_dimension_expansion=True
+    )
+
+    if len(shape) == x.ndim:
+        return Array(x.dask, x.name, x.chunks, meta=x)
```

Comment:
This change would
* fail if chunks were passed and `len(shape) == x.ndim`
* return early before computing `reshape_rechunk` if `len(shape) == x.ndim`
```suggestion
    if chunks is not None:
        raise ValueError(
            "Setting chunks is not allowed when reducing the number of dimensions."
        )

    if len(shape) == x.ndim:
        return Array(x.dask, x.name, x.chunks, meta=x)

    _, _, mapper_in, one_dimensions = reshape_rechunk(
        x.shape, shape, x.chunks, disallow_dimension_expansion=True
    )
```

Reply:
I don't have a strong opinion about the passed chunks case, but we have to keep the order of operations that I added.

Example:

```
# shape (6, 3, 3)
arr = blockwise_reshape(x, (3, 2, 9))
```

This is the same dimensions but it changes the shape, reshape_rechunk raises for cases like this

Reply:
> This is the same dimensions but it changes the shape, reshape_rechunk raises for cases like this

Ah, good point. Let's make sure we raise on passed chunks then as well and this should be good to go.

Reply:
FWIW, I've also just noted that `reshape` raises with a meaningful error when shapes don't match while `reshape_blockwise` is less helpful:

```python
import dask.array as da
arr = da.random.random((10, 10), (2, 2))
da.reshape(arr, (99, ))
```
raises
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/hendrikmakait/projects/dask/dask/dask/array/reshape.py", line 345, in reshape
    raise ValueError("total size of new array must be unchanged")
ValueError: total size of new array must be unchanged```
while 

```python

import dask.array as da
arr = da.random.random((10, 10), (2, 2))
da.reshape_blockwise(arr, (99, ))
```

raises

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/hendrikmakait/projects/dask/dask/dask/array/reshape.py", line 509, in reshape_blockwise
    _, _, mapper_in, one_dimensions = reshape_rechunk(
                                      ^^^^^^^^^^^^^^^^
  File "/Users/hendrikmakait/projects/dask/dask/dask/array/reshape.py", line 68, in reshape_rechunk
    raise NotImplementedError(_not_implemented_message)
NotImplementedError: 
Dask's reshape only supports operations that merge or split existing dimensions
evenly. For example:

>>> x = da.ones((6, 5, 4), chunks=(3, 2, 2))
>>> x.reshape((3, 2, 5, 4))  # supported, splits 6 into 3 & 2
>>> x.reshape((30, 4))       # supported, merges 6 & 5 into 30
>>> x.reshape((4, 5, 6))     # unsupported, existing dimensions split unevenly

To work around this you may call reshape in multiple passes, or (if your data
is small enough) call ``compute`` first and handle reshaping in ``numpy``
directly.
```

It might make sense to add an early check that the shape generally matches.

Reply:
I like that